docs/proposals/event-notification.md

```diff
@@ -0,0 +1,143 @@
+- If a dev file writer wanted to communicate that the current command/action were compiling the application, they would insert a specific markup string (`#devfile-status#`) at the beginning of a console-outputted line, and then between those two fields would be a JSON object with a single field `buildStatus`:
+  - For example: `#devfile-status# {"buildStatus":"Compiling application"}` would then communicate that the detailed build status should be set to `Compiling application`.
+- Since this line would be output as container stdout, it would be included as a `logText` JSON event, and the consuming tool can look for this markup string and parse the simple JSON to extract the detailed build status.
+- Feedback welcome around exact markup text format.
+
+The build step (running as a bash script, for example, invoked via an action) of a devfile might then look like this:
+```
+#!/bin/bash
+(...)
+echo "#devfile-status# {'buildStatus':'Compiling application'}
+mvn compile
+echo "#devfile-status# {'buildStatus':'Running unit tests'}
+mvn test
+```
+
+This 'detailed build status' markup text is *entirely optional*: if this markup is not present, the odo tool can still determine build succeeded/failed and build running/not-running using the other `odo push -o json` JSON events. 
+
+### App status notification via `odo component status -o json` and `odo log --follow`
+
+In general, within the execution context that odo operates, there are a few ways for us to determine the application status:
+1) Can the application be pinged at its exposed URL? 
+2) What state is the container in? (running/container creating/restarting/etc -- different statuses between local and Kube but same general idea)
+3) In the application log, specific hardcoded text strings can be searched for (for example, OpenLiberty outputs defined status codes to its log to indicate that an app started.)
+
+Ideally, we would like for odo to provide consuming tools with all 3 sets of data. Thus, as proposed:
+- 1 and 2 are handled by a new `odo component status -o json` command, described here.
+- 3 is handled by the existing unmodified `odo log --follow` command.
+
+The new proposed `odo component status -o json` command will:
+- Be a *long-running* command that will continue outputing status until it is aborted by the parent process.
+- Every X seconds, ping the URLs/routes of the application as they existed when the command was first executed. Output the result as a JSON string.
+- Every X seconds (or using a Kubernetes watch, where appropriate), check the container status for the application, based on the application data that was present when the command was first issued. Output the result as a JSON string.
+
+**Note**: This command will NOT, by design, respond to route/application changes that occur during or after it is first invoked. It is up to consuming tools to ensure that the `odo component status` command is stopped/restarted as needed. 
```

Comment:
If we do not design it to respond to changes that occur in route/app since first run of `odo component status`, this command would report failure in accessing the route/app, right? For example:
- I pushed the component and executed `odo component status -o json`.
- Now, 
  - I deleted the URL and created another one instead, or
  - I change the port (less likely than URL deletion+creation scenario), or
  - I feel like adding another endpoint to the application `/status` that displays the status of my app.
- If the user has setup some action to be performed in case of an URL/app being inaccessible, wouldn't this give a false positive?

This might be irrelevant right now, but it might become more relevant once #2756 starts moving ahead.

Reply:
Yep, the scenario you describe is exactly the problem with concurrent modification of the application while the status is running, and I included this line to indicate this `Note: This command will NOT, by design, respond to route/application changes that occur during or after it is first invoked. It is up to consuming tools to ensure that the odo component status command is stopped/restarted as needed.` 

How to handle this exact problem is a fundamental architectural constraint in the problem space of this proposal: we need a way to get always up-to-date status, and always up-to-date logs, but we also need the ability to deploy changes to the application. And we need to do this without turning the odo tools into either a server, an LSP-style process, or a multi-process application that needs to do IPC between various running instances. I'll go into more details in your next comment where you have added additional details.

Reply:
I think we agree upon this being a problem since its the end-user who has to perform an action. And as you said elsewhere, this leads to a scenario of the problem being solved at many places instead of being solved once in odo. But that could be taken up in the next phase, I guess.